File: IN/WEEK01/BOJ1865웜홀_TIN.py
import sys
from collections import deque

input = sys.stdin.readline

test = int(input())

for _ in range(test):
    n, m, w = map(int, input().split())
    road = [[] for _ in range(n+1)]
    costs = [1e9] * (n+1)
    costs[1] = 0
    possible = False
    for _ in range(m):
        start, end, time = map(int, input().split())
        road[start].append((time,end))
        road[end].append((time,start))
    for _ in range(w):
        start, end, time = map(int, input().split())
        road[start].append((-time,end))
    for i in range(len(road)):
        road[i].sort()
    for i in range(1,n+1):
        for j in range(1,n+1):
            for cost, next in road[j]:
                if costs[next] > costs[j] + cost:
                    costs[next] = costs[j] + cost
                    if i == n:
                        possible = True
                
    if possible:
        print("YES")
    else:
        print("NO")
        

File: IN/WEEK11/BOJ1865웜홀_TIN.cpp
#include <iostream> 
#include <vector>
#include <algorithm>
#include <utility>
#include <queue>
using namespace std;

vector< vector< pair<int, int> > > road;
vector<int> costs;

int INF = int(1e9);

int main(void){
    int test, n, m, w;
    scanf("%d", &test);
    for(int i = 0; i < test; ++i){
        scanf("%d %d %d", &n, &m, &w);
        road = vector< vector< pair<int, int> > >(n+1);
        costs = vector<int>(n+1, INF);
        costs[1] = 0;
        bool possible = false;
        int start, end, time;
        for(int j = 0; j < m; ++j){
            scanf("%d %d %d", &start, &end, &time);
            road[start].push_back(make_pair(time, end));
            road[end].push_back(make_pair(time, start));
        }
        for(int j = 0; j < w; ++j){
            scanf("%d %d %d", &start, &end, &time);
            road[start].push_back(make_pair(-time, end));
        }
        for(int j = 0; j < road.size(); ++j){
            sort(road[j].begin(), road[j].end());
        }
        for(int j = 1; j <= n; ++j){
            for(int k = 1; k <= n; ++k){
                for(int z = 0; z < road[k].size(); ++z){
                    pair<int, int> next = road[k][z];
                    if(costs[next.second] > costs[k] + next.first){
                        costs[next.second] = costs[k] + next.first;
                        if(j == n) possible = true;
                    }
                }
            }
        }
        if(possible) printf("YES\n");
        else printf("NO\n");
    }

    return 0;
}
